Add draft yoastseo package overview (#21876)
Yoast/wordpress-seo: packages/yoastseo/GLOSSARY.md

# YoastSEO.js Glossary

A comprehensive glossary of terms and concepts used in YoastSEO.js.

## Core Concepts

### <a name="paper"></a>Paper
A value object that encapsulates all content and metadata to be analyzed. The Paper object is immutable and serves as the primary input for all analyses.

**Properties include:**

> - text: The main content to analyze

Reviewer: ```suggestion
- text: The main content to analyze. We assume the content follows a standard HTML format, and consists of the elements that would be part of the `<body>` (so no `<head>` or `<footer>` elements.
```

> - title: The SEO title or page title

Reviewer: ```suggestion
- title: The SEO title
```

- keyword: The focus keyphrase
- description: Meta description

> - url: The slug/URL of the page

Reviewer: ```suggestion
- slug: The slug/URL of the page
```

- locale: Language code (e.g., en_US)
- permalink: Full URL

**Example:**
```javascript

> const paper = new Paper("This is the main content", {

Reviewer: ```suggestion
const paper = new Paper("<p>This is the <strong>main</strong> content<p>", {
```

    title: "Example Title",
    keyword: "example",
    description: "This is a meta description",

>     url: "example-page",

Reviewer: ```suggestion
    slug: "example-page",
```

    locale: "en_US"
});
```

### <a name="assessment"></a>Assessment
A single analysis unit that evaluates one specific aspect of content. Each assessment:

> - Has a specific purpose (e.g., keyword density analysis)

Reviewer: ```suggestion
- Has a specific purpose (e.g., the _keyword density_ assessment evaluates the number of keywords used in the content)
```

> - Produces a score (0-100)

Reviewer: ```suggestion
- Produces a score (0-9)
```

Author: Is that meant to be `0-90`?

Reviewer: No `0-9` is actually correct 😄

Reviewer: (well, in extreme cases, you can have negative scores as well, e.g. if you do not enter a focus keyphrase)

> - Can generate improvement suggestions
> - May include text markers for visual feedback

Reviewer: ```suggestion
- May include text markers for visual feedback
- Can generate improvement suggestions (through Yoast AI)
```

**Example Assessment Structure:**
```javascript
class KeywordDensityAssessment extends Assessment {
    getResult(paper, researcher) {
        const density = researcher.getResearch("keywordDensity");
        return {
            score: this.calculateScore(density),
            text: this.translateScore(density)
        };
    }
}
```

### <a name="assessor"></a>Assessor
A collection manager that:
- Coordinates multiple assessments
- Determines which assessments to run
- Aggregates assessment results
- Calculates overall scores

Types of assessors include:
- SEOAssessor: Focuses on search engine optimization
- ReadabilityAssessor: Analyzes text readability
- CornerStoneAssessor: Applies stricter rules for important content

> ```mermaid

Reviewer: ```suggestion
The diagram below shows an example hierarchy of assessors and assessments. 

```mermaid
```

graph TD
    A[Assessor] --> B[SEOAssessor]
    A --> C[ReadabilityAssessor]
    A --> D[CornerStoneAssessor]
    B --> E[KeywordAssessment]
    B --> F[URLAssessment]
    C --> G[SentenceLengthAssessment]
    C --> H[ParagraphLengthAssessment]
```

### <a name="researcher"></a>Researcher
The research component that:
- Performs linguistic analysis
- Caches research results
- Provides research data to assessments

Common research types:
- Word count
- Sentence detection
- Keyword presence
- Morphological analysis

**Example Research Usage:**
```javascript
const researcher = new Researcher(paper);
const wordCount = researcher.getResearch("wordCountInText");
const sentences = researcher.getResearch("sentences");
```

### <a name="score"></a>Score
A numeric representation (0-100) of content quality for a specific aspect.

Score ranges:

> - 0-40: Bad (Red)
> - 41-70: Needs Improvement (Orange)
> - 71-100: Good (Green)

Reviewer: ```suggestion
- 0-30: Bad (Red)
- 31-60: Needs Improvement (Orange)
- 61-90: Good (Green)
```

Author: @mhkuu There seems to be some confusion somewhere around the scoring, and I'm not sure what to rely on.
Here's one of the spec tests:
https://github.com/Yoast/wordpress-seo/blob/5d7ffe3f1b99c156ae606d871c994934bbdd1bd4/packages/yoastseo/spec/scoring/renderers/AssessorPresenterSpec.js#L34-L46
Where do you get the adapted scale from?

### <a name="marker"></a>Marker
A system to highlight relevant parts of text for specific assessments. Markers:
- Help visualize assessment results
- Provide inline feedback

> - Support multiple marking types (error, warning, good)

Reviewer: ```suggestion
```

### <a name="morphology"></a>Morphology
The study of word forms and their variations. In YoastSEO.js, morphology is used to:
- Recognize different forms of keywords
- Improve keyword matching
- Support language-specific word variations

**Example:**
```
Base word: teach
Morphological forms: teaches, teaching, taught, teacher, teachers
```

### <a name="stem"></a>Stem
The base form of a word before any affixes. Stemming helps in:
- Keyword matching
- Word form recognition
- Content analysis

**Example:**
```
Word: running
Stem: run
Related forms: runs, ran
```

### <a name="function-words"></a>Function Words
Words that serve grammatical purposes but carry little meaning:
- Articles (the, a, an)
- Prepositions (in, on, at)
- Conjunctions (and, or, but)
- Auxiliary verbs (is, has, will)

These are often filtered out during analysis to focus on meaningful content.

### <a name="content-words"></a>Content Words
Words that carry semantic meaning:
- Nouns (book, house)
- Main verbs (run, write)
- Adjectives (big, red)
- Adverbs (quickly, well)

These are crucial for:
- Keyword analysis
- Topic detection
- Content quality assessment

### <a name="keyphrase"></a>Keyphrase

> The main search term or topic being targeted. Can be:
> - Single word ("SEO")
> - Multiple words ("WordPress SEO plugin")
> - Contains function words ("how to bake bread")

Reviewer: ```suggestion
The main search term or topic being targeted. We use `keyphrase` and `keyword` interchangeably, but primarily the former, as a keyphrase can be a single word ("SEO") or consist of multiple words ("WordPress SEO plugin"), and even contain function words ("how to bake bread").
```
